react/src/pages/admin/UserDetails.tsx

```diff
@@ -0,0 +1,204 @@
+import SecretDisplay from "../utils/components/SecretDisplay";
+import NewPasswordForm, { ConfirmPasswordAction } from "../utils/components/NewPasswordForm";
+import ChipTransferList, { TransferChip } from "../utils/components/ChipTransferList";
+import ConfirmModal from "./ConfirmModal";
+
+const useDetailStyles = makeStyles(theme => ({
+    root: {
+        margin: theme.spacing(1),
+        padding: theme.spacing(2),
+    },
+    grow: {
+        flexGrow: 1,
+    },
+    toolbar: {
+        padding: 0,
+    },
+    button: {
+        marginRight: theme.spacing(1),
+    },
+}));
+
+function reducer(state: User, action: ConfirmPasswordAction | { type: "set"; payload: User }) {
+    switch (action.type) {
+        case "password":
+            return { ...state, password: action.payload };
+        case "confirm":
+            return { ...state, confirmPassword: action.payload };
+        case "set":
+            return { ...action.payload };
+        default:
+            return state;
+    }
+}
+
+/**
+ * The collapsible part of a user row. A form for viewing
+ * and editing credentials.
+ */
+export default function UserDetails(props: {
+    user: User;
+    onSave: (newUser: User) => void;
+    onDelete: (deleteUser: User) => void;
+}) {
+    const classes = useDetailStyles();
+    // Local changes saved in newState, and are "committed" when user saves changes
+    const oldState = {
+        ...props.user,
+        password: "",
+        confirmPassword: "",
+        groupMemberships: ["FOO", "BAR"],
+    } as User;
+    const [newState, dispatch] = useReducer(reducer, oldState);
+
+    const [confirmDelete, setConfirmDelete] = useState(false);
+    const [confirmSave, setConfirmSave] = useState(false);
+
+    // TODO: hook up with actual permission groups
+    const [tempGroups, setTempGroups] = useState<TransferChip[]>(
+        ["FOO", "BAR", "BAZ"].map((label, index) => ({
+            label: label,
+            key: index,
+            selected: false,
+        }))
+    );
+
+    return (
+        <>
+            <ConfirmModal
+                id="confirm-modal-update"
+                color="primary"
+                open={confirmSave}
+                onClose={() => setConfirmSave(false)}
+                onConfirm={() => props.onSave(newState)}
+                title="Confirm updating user"
+            >
+                Are you sure you want to save changes to user {oldState.username}?
+            </ConfirmModal>
+            <ConfirmModal
+                id="confirm-modal-delete"
+                color="secondary"
+                open={confirmDelete}
+                onClose={() => setConfirmDelete(false)}
+                onConfirm={() => props.onDelete(oldState)}
+                title="Delete user"
+            >
+                Are you sure you want to delete user {oldState.username}?
+            </ConfirmModal>
+            <Box className={classes.root}>
+                <Grid container xs={12} spacing={1}>
```

Comment:
```suggestion
                <Grid container spacing={1}>
```
Missed this one more case of `xs` on a non-item